devito/types/basic.py

```diff
@@ -1014,29 +1037,31 @@ def _evaluate(self, **kwargs):
         This allow to evaluate off grid points as EvalDerivative that are better
         for the compiler.
         """
+        mapper = self._grid_map
+        subs = mapper.pop('subs', {})
         # Average values if at a location not on the Function's grid
-        if not self._grid_map:
+        if not mapper:
             return self
 
         io = self.interp_order
-        # Base function
-        if self._avg_mode == 'harmonic':
-            retval = 1 / self.function
-        else:
-            retval = self.function
+        retval = self.subs({i.subs(subs): self.indices_ref[d]
+                            for d, i in mapper.items()})
+        if 'harmonic' in self._avg_mode:
+            retval = retval.safe_inv(retval, safe='safe' in self._avg_mode)
 
         # Apply interpolation from inner most dim
-        for d, i in self._grid_map.items():
+        for d, i in mapper.items():
             retval = retval.diff(d, deriv_order=0, fd_order=io, x0={d: i})
 
         # Evaluate. Since we used `self.function` it will be on the grid when
         # evaluate is called again within FD
         retval = retval._evaluate(**kwargs)
+        retval = retval.subs(subs)
 
         # If harmonic averaging, invert at the end
-        if self._avg_mode == 'harmonic':
-            from devito.finite_differences.differentiable import SafeInv
-            retval = SafeInv(retval, self.function)
+        if 'harmonic' in self._avg_mode:
```

Comment:
nitpicking: the `_avg_mode` should be an instance of something rather than an iterable of string, so that you may do e.g. `self._avg_mode.is_harmonic` or `self._avg_mode.is_harmonic_safe`